### ISSUE TITLE
Encrypt local password

### ISSUE THREAD

**Issue by superyuuki** [vrtmrz/obsidian-livesync#59]
I noticed that the password is stored in plaintext in the config file locally. Can you set up some kind of encryption?

**Comment by vrtmrz** [comment]
@superyuuki Thank you for investigating LiveSync deeply!

I was wondering how I can implement an effective method to protect our vault.

May I ask what the purpose is?

(Bit long)

If we store the passphrase or authorization data into `data.json` with some encryption, we have to decrypt them to use in the plugin. The plugin will use the logic in this repo, so it won't protect you totally while they know it's "Self-hosted LiveSync".
(When the attacker couldn't infer the plugin's name, encryption will protect you some much.)

If we store the secrets into Obsidian's configuration instead of storing them into `data.json`, It will protect you, but you can't back your vault up completely or switch `.obsidian` easily.

If you want protection just from casual peeking, it's very easy. 
As like `supercow` -> `==c3VwZXJjb3c=`.
We can decode this easily.

**Comment by vrtmrz** [comment]
Feel free to open it again if you have new something.

**Comment by hilsonp** [comment]
Could we re-open this please ?
The data.json not only has one password.
It has:
- couchDB_URI
- couchDB_USER
- couchDB_PASSWORD
- couchDB_DBNAME
- passphrase
- workingPassphrase

# Passphrase

I thought the passphrase was use to:
- encode the URI generated by the "Copy setup URI"
- decode the URI when using the "Open setup URI"

Then, why is there a need to store it ? (I guess, this is so you can persist the "Remote Database configuration / Passphrase" field and generate several time the same URI without asking the passphrase several time)

I think, as a user, I should store this URI and the associated passphrase in a password manager.
Or just not store it and generate a new URI with a new passphrase (or the same that I remembered) then use that URI immediately to setup a new device.
If I need to setup another device later on, I can generate a new URI with yet another passphrase and use that new one on the "another device".

# DB connection data

Concerning the couchDB_URI, couchDB_USER, couchDB_DBNAME, I understand you need them.
I also understand that you need to need somehow to store the couchDB_PASSWORD but could you please at least offuscate it so it is not displayed in clear text to pry eyes would I need to open this file ?
[This is rather interresting](https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html) although, as said, offuscation would already be nice and 1000x better than now.

Thank you,

Pierre

**Comment by vrtmrz** [comment]
> Could we re-open this please ?

Of course! I will write my response soon.

**Comment by vrtmrz** [comment]
I am sorry that I didn't write definitions for these terms. I used `passphrase` for the term for encrypting or decrypting something. (`Password` is for the authentication)

Let me explain what I have thought and currently think. A bit long.

The `passphrase` which is stored in `setting.json` and could be configured in the setting dialogue is for encrypting and decrypting notes on the remote database. All content of notes has been encrypted with this and transferred. The other passphrase that we are asked for, and will answer when using the Setup URI is for encrypting and decrypting the Setup URI itself. 
So, we have two passphrases, one to secure the remote vault and another to secure the Setup URI. We may need to name each of them. For the moment, the `E2EE passphrase` and `URI passphrase`.

The `URI passphrase` will not be stored anywhere for safekeeping. Therefore, we should store it with the Setup URI or should be determined for each Setup URI. You are absolutely right. Thank you for your good point. And I know that you had already realised, if it leaks with the Setup URI, anyone can synchronise with our vault instantly. The Setup URI contains every other password and passphrase.

In contrast, `E2EE passphrase` and `Password` should also be kept secret. However, we have to share them with every device and they used every synchronising. 

I have thought that it is only helpful for limited scenes because we could look at them in memory, even if they were stored obfuscated in the storage. But now I think we might be able to use it at least we can use it a little bit more safely while we are off the line. 

As you know, the implementation of this plug-in is open, anyone can see how to obfuscate it. Then, anyone can decode it. However, as you said, not all users check, do they? Therefore, I think that it is a good idea to obfuscate.

Only one that we have to keep in mind; Actually, nothing has changed. It has not really become safer. It may sound disgusting but there can be a decoder if evil wants. In that meaning, even though it is obfuscated, it should be treated the same as the plain. If they are not encrypted by something which not being stored in Obsidian.

Then, to solve this problem completely; a new feature that really encrypts them and asks before every launch or synchronisation would be needed.

Thank you for reading the long and complicated note! I got inspired as I said last.

P.S. workingXXX is a bit harsh. It might be out of the question and would be fixed.

**Comment by hilsonp** [comment]
Thank you for the complete explanation. This is really clear. I forgot, or did not understand, the encryption made when storing the notes inside the database. 

I think you can reuse as-is in a FAQ. 

Obfuscation would be a good think. 

Not everyone would go read the code. And if someone is going to go read the code, then, he is probably skilled enough to hack my system anyway. 

My vault does not store sensitive information. Those are stored in a self-hosted password manager 😉

I just don't like to have passwords in clear text in configurations files or logs. 

This is not urgent thought. 
In the mean time, I will redo the setup with much stronger passwords that no one can remember by just looking at them 😁

**Comment by vrtmrz** [comment]
I am very happy to hear that!

The faq might be an incredible idea. I will make them.

> I just don't like to have passwords in clear text in configurations files or logs.
 
I am feel so too. That is also why the log of LiveSync is volatile. Might be a nightmare if the confidential information is dripping down without our control.

> This is not urgent thought.

Thankfully, writing notes made my ideas clear. I tried an implementation and it could be made up straightforwardly. 
It has been released at v0.17.15. I will be very happy to hear your impressions.

I would like to express my appreciation again!
